src/lib/core/cbus.c

```diff
@@ -177,6 +187,145 @@ cpipe_destroy(struct cpipe *pipe)
 	TRASH(pipe);
 }
 
+struct lcpipe *
+lcpipe_new(const char *consumer)
+{
+	struct lcpipe *pipe = xmalloc(sizeof(*pipe));
+
+	stailq_create(&pipe->input);
+
+	pipe->n_input = 0;
+	pipe->max_input = INT_MAX;
+
+	acquire_consumer(&pipe->endpoint, consumer);
+
+	return pipe;
+}
+
+/**
+ * Flush all staged messages into consumer output and wake up it after.
+ */
+static void
+lcpipe_flush(struct lcpipe *pipe)
+{
+	struct cbus_endpoint *endpoint = pipe->endpoint;
+	if (pipe->n_input == 0)
+		return;
+
+	/* Trigger task processing when the queue becomes non-empty. */
+	bool output_was_empty;
+
+	/*
+	 * We need to set a thread cancellation guard, because
+	 * another thread may cancel the current thread
+	 * (write() is a cancellation point in ev_async_send)
+	 * and the activation of the ev_async watcher
+	 * through ev_async_send will fail.
+	 */
+
+	int old_cancel_state;
+	tt_pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_cancel_state);
+
+	tt_pthread_mutex_lock(&endpoint->mutex);
+	output_was_empty = stailq_empty(&endpoint->output);
+	/** Flush input */
+	stailq_concat(&endpoint->output, &pipe->input);
+	tt_pthread_mutex_unlock(&endpoint->mutex);
+
+	pipe->n_input = 0;
+	if (output_was_empty) {
+		/* Count statistics */
+		rmean_collect(cbus.stats, CBUS_STAT_EVENTS, 1);
+
+		ev_async_send(pipe->endpoint->consumer, &pipe->endpoint->async);
+	}
+
+	tt_pthread_setcancelstate(old_cancel_state, NULL);
+}
+
+void
+lcpipe_flush_input(struct lcpipe *pipe)
+{
+	/** Flush may be called with no input. */
+	if (pipe->n_input > 0) {
+		lcpipe_flush(pipe);
+	}
+}
+
```

Comment:
`lcpipe_flush` already checks `n_input` and doesn't do anything when it is zero. This check is not needed. Which makes `lcpipe_flush_input` == `lcpipe_flush`, which means you can rename `lcpipe_flush` to `lcpipe_flush_input` and delete this code block.

Reply:
thanks, done